plugin/templates/resource_name.py

```diff
@@ -188,3 +289,54 @@ def __init__(self, fixed_config, java_package, oneof):
 
     def template_name(self):
         return "resource_name_fixed.mustache"
+
+
+def get_id_segments(pattern):
+    name_template = path_template.PathTemplate(pattern)
+    return [
+        seg.literal for seg in name_template.segments
+        if seg.kind == path_template._BINDING
+    ]
+
+
+def get_format_field(lower_underscore, symbol):
+    return {
+        'lower_underscore': lower_underscore,
+        'lower_camel': casing_utils.lower_underscore_to_lower_camel(
+                       lower_underscore),
+        'lower_camel_symbol': symbol,
+        'upper_underscore': casing_utils.lower_underscore_to_upper_underscore(
+                            lower_underscore),
+        'upper_camel': casing_utils.lower_underscore_to_upper_camel(
+                       lower_underscore),
+        'not_first': True,
+        'not_last': True
+    }
+
+
+def get_format_fields_for_pattern(pattern,
+                                  pattern_to_id_segments,
+                                  segment_to_segment_symbols):
+    if pattern not in pattern_to_id_segments:
+        return []
+    format_fields = [get_format_field(seg, segment_to_segment_symbols[seg])
+                     for seg in pattern_to_id_segments[pattern]]
+    format_fields[0]['not_first'] = False
+    format_fields[-1]['not_last'] = False
+    return format_fields
+
+
+def is_fixed_pattern(pattern):
+    return not('{' in pattern or '*' in pattern)
+
+
+def get_pattern_name(pattern):
+    if is_fixed_pattern(pattern):
+        start_index = \
+            next(i for (i, c) in enumerate(list(pattern)) if c.isalpha())
+        end_index = len(pattern) - next(
+            i for (i, c) in enumerate(list(pattern)[::-1]) if c.isalpha())
```

Comment:
`[::-1]` - wow, this is how you reverse an array... 

Looks very neat, but also looks like same thing can be achieved in a more efficient, but maybe less fancy way by simply iterating by the characters of the string (using one loop to get both `start_index` and `end_index`).

It is not a big deal here, but please pay attention to how much inefficiency was introduced here:
1) first convert a string to an array (`list());
2) enumerate it (probably very cheap, though);
3) do all the above twice.

and it is simply to get first and last letter character.

Reply:
This method is copied from `gapic_utils.py` ([here](https://github.com/googleapis/protoc-java-resource-names-plugin/blob/2e002340bb689744adf0e537617a3885155ee024/plugin/utils/gapic_utils.py#L390))

I can't import it from the other file because it will cause circular dependency...